### PR TITLE
Add split and retry support for filter [databricks]

### PR DESCRIPTION
**Pull request by revans2**
I was doing some benchmarking and needed a tiered project in filter to make my test faster. When looking at it I noticed that we didn't have retry for filter yet so I added it in.  I also cleaned up some of GpuProject to hide values from the user, like tiered project, that they should not care about. We can still see if tiered project is enabled by looking at the code and the configs.

### CODE REVIEW

**Comment by revans2** [comment]
build

**Review by abellina — DISMISSED** [review]

**Comment by abellina** [comment]
Looks like one of the currying changes needs to be made in 340:
```
sql-plugin/src/main/spark340/scala/org/apache/spark/sql/rapids/GpuFileFormatWriter.scala:226: missing argument list for method apply in object GpuProjectExec
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala [hunk 1]
```diff
@@ -448,6 +452,33 @@ object GpuFilter {
     }
   }
 
+  def filterAndCloseWithRetry(input: SpillableColumnarBatch,
+      boundCondition: GpuTieredProject,
+      numOutputRows: GpuMetric,
+      numOutputBatches: GpuMetric,
+      opTime: GpuMetric): Iterator[ColumnarBatch] = {
+    val ret = withRetry(input, splitSpillableInHalfByRows) { sb =>
```
Comment (jbrennan333) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:460]: Do we need to avoid retries if there are non-deterministic expressions in boundCondition?

**Review by jbrennan333 — COMMENTED** [review]
This looks good. Just one question.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala (hunk 1 above)
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:460]: You are right. Oops...

**Comment by revans2** [comment]
Turns out we did support retry (but split and retry we didn't support, so I am adding that here).

**Comment by revans2** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala [hunk 2]
```diff
@@ -434,20 +438,48 @@ object GpuFilter {
     }
   }
 
-  def filterAndClose(
-      batch: ColumnarBatch,
+  def filterAndCloseNondeterministic(batch: ColumnarBatch,
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:441]: I don't see this function getting called as of HEAD

**Comment by revans2** [comment]
Thanks for your patience @abellina . I should have done more testing/work up front. I think I have it all working now. I ran filter tests locally with the error injection and I added a test to verify that the non-deterministic case is covered, so if we do attempt to retry it we would get test failures. Please take another look

**Comment by revans2** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala [hunk 3]
```diff
@@ -488,20 +527,15 @@ object GpuFilter {
...
-      }
-    } else {
-      val sb = SpillableColumnarBatch(batch, SpillPriorities.ACTIVE_ON_DECK_PRIORITY)
-      RmmRapidsRetryIterator.withRetryNoSplit(sb) { sb =>
-        withResource(sb.getColumnarBatch()) { cb =>
-          GpuFilter(cb, boundCondition)
-        }
+  private def computeCheckedFilterMask(boundCondition: GpuTieredProject,
+      cb: ColumnarBatch): Option[cudf.ColumnVector] = {
+    withResource(boundCondition.project(cb)) { filterBatch =>
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:532]: This not using `projectSingle` because there isn't one for tiered project I imagine. Might be nice to add one in a different PR

Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:532]: Also, I can't think of a case where the projection would result in 0 columns, but bringing it up in case we need to be defensive.

Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:532]: There are cases when it could COUNT is one where all we have is rows, no columns. I could add in a check to verify that there is one and only one output column. I think it would be `boundExprs.tail.length == 1`, but I didn't want to spend a ton of time on it.

**Review by abellina — DISMISSED** [review]

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 4]
```diff
@@ -3803,8 +3803,10 @@ object GpuOverrides extends Logging {
           TypeSig.ARRAY + TypeSig.DECIMAL_128 + TypeSig.BINARY +
           GpuTypeShims.additionalCommonOperatorSupportedTypes).nested(), TypeSig.all),
       (filter, conf, p, r) => new SparkPlanMeta[FilterExec](filter, conf, p, r) {
-        override def convertToGpu(): GpuExec =
-          GpuFilterExec(childExprs.head.convertToGpu(), childPlans.head.convertIfNeeded())
+        override def convertToGpu(): GpuExec = {
+          GpuFilterExec(childExprs.head.convertToGpu(),
+            childPlans.head.convertIfNeeded())(conf.isTieredProjectEnabled)
```
Comment (jbrennan333) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:3808]: (nit) Should we use the named parameter `useTieredProject` here, and in the few other cases where we pass it, and not rely on the order of the two boolean paramaters?

**Review by jbrennan333 — COMMENTED** [review]
Looks good.
I just had the one additional nit (and it looks like there is a build failure).

**Comment by revans2** [comment]
build

**Review by jbrennan333 — APPROVED** [review]
